Add PQL support, refactor and reorganize code

gdubicki opened the pull request:
~~Basic~~ Not so basic anymore ~~but~~ and working PQL support. ;)

There are 2 ways of using PQL:
* as a type of query for existing methods, like `nodes()` - this gives you casting to our rich types like `Node`, but limits ability to use projections,
* as a direct call to the `pql` endpoint - this gives all the features of PQL but returns raw JSON = list of dicts,

While working on this I revamped the user docs.

While doing that I have added info on how to use the `ssl_verify` to provide the CA root certificate (fixes #143) and added missing info about using the Basic auth. I also improved some examples, fixing #129.

*UPDATE*: the scope and API have changed since I created this PR, see the comments below.

gdubicki commented on docs/advanced_queries.rst:
```diff
@@ -0,0 +1,206 @@
+   >>> query = SubqueryOperator('facts')
+   >>> query.add_query(op)
+   >>> print(query)
+   >>> # TODO: fix above code as this is not the query we wanted :/
```
I need some help here as I can't recreate the AST query from the section above using the QueryBuilder syntax... (Maybe that's the reason I am working on this PQL PR..? ;P )

b4ldr replied:
I searched the code and as far as i can tell the query builder never uses subquery directly and uses explicit subqueries using select_*.  I think the folllowing may represent the same query

```python
    >>> op = InOperator('certname')
    >>> query = ExtractOperator()
    >>> query.add_field(str('certname'))
    >>> subquery = SubqueryOperator('facts')
    >>> _add = AndOperator()
    >>> _add.add(EqualsOperator('name', 'operatingsystem'))
    >>> _add.add(EqualsOperator('value', 'Debian'))
    >>> subquery.add_query(_add)
    >>> query.add_query(subquery)
    >>> op.add_query(query)
    >>> print(op)
   ["in", "certname", ["extract", ["certname"], ["select_facts", ["and", ["=", "name", "operatingsystem"], ["=", "value", "Debian"]]]]]
```

gdubicki replied:
Great, thanks!

b4ldr reviewed (commented):
Im no expect on pypuppetdb but this looks like a great addition to me thanks

gdubicki commented:
Frankly I am not very happy with the proposed API design. We would be repeating ourselves when using PQL and rich types.

```python
db.nodes(pql="nodes { (...) })
#  ^^^^^      ^^^^^   
#      \ ugly /
```

I don't want to mangle with the PQL query as this will creating it / testing it elsewhere (f.e. in Query view of the Puppetboard) harder.

So this leaves us with changing the method name. Maybe we should do autocasting when projections are not used?

```python
# this would return Nodes
db.pql("nodes { (...) }")

# this would return Inventories
db.pql("inventory { (...) }")
```

...with automatic projection detection:

```python
# this would return dicts as projection is being used
db.pql("inventory[foobar] { (...) }")
```

...and an optional extra parameter to disable the casting:

```python
# this would return dicts
db.pql("nodes { (...) }", raw=True)
```

But then I would NOT add the `pql` parameter support to existing methods like `nodes()` - as "there should be only way to do it". 🧘‍♂️

Thoughts?

bastelfreak commented:
I don't know enough Python to give a proper statement here. Maybe @sebastianrakel is able to review this.

b4ldr reviewed (commented):
> ```python
> db.nodes(pql="nodes { (...) })
> #  ^^^^^      ^^^^^   
> #      \ ugly /
> ```

I also thought this looked a bit strange however ...

> To remove this duplication we could ask the user to strip PQL of the "nodes {" and "}" parts but I think that would be wrong. 

I thought this was worse (and so didn't comment)


> How about we autocast when projections are not used, like this:
> ...and an optional extra parameter to disable the autocasting completely:

I like the idea of auto casting however i think the raw pql method and the autocasting method should be two separate methods

 ```python
def _pql(self, pql, **kwargs):
    elements = self._query(pql=pql, **kwargs)
        for element in elements:
            yield element

def pql(self, pql, **kwargs):
      for element in _pql(pql, **kwargs):
         try:
             yield # cast element
        except:
             yield element
```
Further as to my comments in the review i think it could make sense to change the code so that the `_pql` method dosn;t rely on `_query` instead split out the logic to make the actual request into its own method `_request` and have _query and _pql both call that  

> But then I would NOT add the `pql` parameter to existing methods like `nodes()` as "there should be only way to do it". lotus_position_man
agreed

gdubicki commented:
Thanks a lot for the insightful review, @b4ldr ! I will try to apply your suggestions within the next few days.

gdubicki commented on pypuppetdb/api.py:
```diff
@@ -1022,6 +1047,24 @@ def inventory(self, **kwargs):
                 trusted=inv['trusted']
             )
 
+    def pql(self, pql, **kwargs):
```
Done.

gdubicki commented on pypuppetdb/api.py:
```diff
@@ -337,60 +338,80 @@ def _query(self, endpoint, path=None, query=None,
+            raise APIError
+
+        pql_unsupported_args = ['order_by', 'limit', 'include_total', 'offset', 'summarize_by',
+                                'count_by', 'count_filter']
```
Pulled `_request()` as a separate method.

gdubicki commented:
I have applied your suggestions about the code, @b4ldr (but not the docs yet) and while doing that I got carried away a bit...

I have refactored the code a lot.

Firstly I have splitted the `BaseAPI` from a single > 1000 LOC long `api.py` into separate classes and files by the query endpoints, in a way almost exactly like they are listed in the [PuppetDB docs](https://puppet.com/docs/puppetdb/latest/index.html):
* `pdb/query/v4` - `PqlAPI`
* `pdb/query/v4/*` - `QueryAPI`
* `pdb/cmd/v1` - `CommandAPI`
* `status/v1/**` - `StatusAPI`
* `pdb/meta/v1*` - `MetadataAPI`
* `metrics/*` - `MetricsAPI`

(I have done a similar split in the tests.)

They all inherit from `BaseAPI` where I left the code shared by all the APIs and I used multiple inheritance to combine those APIs into a single `API` class that should be used externally.

Secondly have I pulled out the methods to create the entities like `Node` into those classes, as `create_from_dict()` methods.

I am sure that this is better than what we had before, but not sure if this is the perfect design. What do you think, @b4ldr ?

b4ldr commented:
Just a quick note to say this is on my list and thanks for all the effort, hope to give it a review over the weekend

bastelfreak commented:
hey, this diff is quite big. Is it possible to split it into multiple PRs? that makes it easier to review and better for a generated changelog.

gdubicki commented:
Yes, I know that the PR is big. I don't like such PRs too, but IMHO this code badly needed refactoring.

So there are 3 groups of changes here: 1. PQL support added, 2. Reorganization and very small refactoring of the rest of the code, 3. Docs update.

The refactoring in 2. is minimal to keep the PR manageable. I just pulled some code out into separate methods and did not change the existing logic, with one exception: I fixed #202 . I changed those 2 lines mentioned in that issue. So most of 2. is moving the code around.

I can pretty easily get the 3. out of this PR. Would that be enough, @bastelfreak ?

bastelfreak commented:
getting 3 out would be a good start. can you do that, we review and merge the new PR and afterwards rebase #201 ?

gdubicki commented:
I moved docs update out of this PR but I think that we have a misunderstanding about that, @bastelfreak .

Docs update is related to the other changes here as not only it add the docs for the new PQL API but also puts PQL in the first place as the recommended way of querying PuppetDB.

So IMHO we should review this PR first and _then_ then I'll create the PR with docs.

b4ldr commented on pypuppetdb/api/pql.py:
```diff
@@ -0,0 +1,152 @@
+        return self._make_request(url, request_method, payload)
+
+    # TODO: deduplicate this - see QueryAPI.nodes()
+    def pql(self, pql, with_status=False, unreported=2, with_event_numbers=True):
```
don't feel right to have node specific parameters for such a generic pql method.

gdubicki replied:
Same here, but I don't have a better idea..

b4ldr reviewed (commented):
I think this looks good to me as you mentioned most of the changes are just moving code around.  Im not sure i like the big mix-in API class however i think the organisation is much better.  however it would be good to try all of the mix-in classes implement unique methods.  i,e. some way to prevent other class implementing `query` 

The main risk i see is if someone is using BaseAPI directly which seems unlikely to me but still worth highlighting.  I think it would be very good and increase my confidence if we could show that this new version works with puppetboard (with no alterations).  Is this something you are able to test?

gdubicki commented:
> I think this looks good to me as you mentioned most of the changes are just moving code around. Im not sure i like the big mix-in API class however i think the organisation is much better. however it would be good to try all of the mix-in classes implement unique methods. i,e. some way to prevent other class implementing `query`

Frankly I don't think that we should implement protection from something like this - this problem should be caught by the tests or in a code review. 

> The main risk i see is if someone is using BaseAPI directly which seems unlikely to me but still worth highlighting. I think it would be very good and increase my confidence if we could show that this new version works with puppetboard (with no alterations). Is this something you are able to test?

I did such test and found only one bug that I fixed in 9e3048f. (I also splitted `other.py` into a separate file for each class - better to do it now, that another reorg in the future.)

If we get this merged then I propose to create a pre-release of pypuppetdb and a pre-release of puppetboard depending on it. Then everyone interested (like me :)) can test it. Only after we won't find bugs after some time we would release a final release. What do you think?

b4ldr commented:
> Frankly I don't think that we should implement protection from something like this - this problem should be caught by the tests or in a code review.

Perhaps my language was to terse i meant protections as in unit test, definitely not some type of run time check (not even sure how that would look) so i think we agree here.

> I did such test and found only one bug that I fixed in [9e3048f](https://github.com/voxpupuli/pypuppetdb/commit/9e3048fb40ffdb8ca618e7317579f6bd6256aa7a). (I also splitted `other.py` into a separate file for each class - better to do it now, that another reorg in the future.)

Awesome <3

> If we get this merged then I propose to create a pre-release of pypuppetdb and a pre-release of puppetboard depending on it. Then everyone interested (like me :)) can test it. Only after we won't find bugs after some time we would release a final release. What do you think?

This sound like a good plane to me

sebastianrakel reviewed (commented):
The code looks very clean, but i'm missing a test TestVersionCmp. Is this right?

gdubicki commented:
> The code looks very clean, but i'm missing a test TestVersionCmp. Is this right?

Oh, I forgot that I removed `versioncmp()` because it was unused here. But you made me realize that it may be used in someone's code using this library. I'll just restore it as I don't want to break anyone's code without a good reason.

gdubicki commented:
I cleaned up the history but I didn't squash everything as I think that some development history kept in the separate commits and their messages may be useful.

@bastelfreak : you are the 3rd reviewer, are you ok with emerging this?

gdubicki commented:
A draft of the PR for updating the docs to add PQL and the new code structure is here: https://github.com/gdubicki/pypuppetdb/pull/1
